functions/snippets/src/main/java/com/example/functions/OcrProcessImage.java

```diff
@@ -0,0 +1,152 @@
+import com.google.cloud.vision.v1.ImageSource;
+import com.google.gson.Gson;
+import com.google.protobuf.ByteString;
+import com.google.pubsub.v1.ProjectTopicName;
+import com.google.pubsub.v1.PubsubMessage;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.ExecutionException;
+import java.util.logging.Logger;
+
+public class OcrProcessImage implements BackgroundFunction<GcsEvent> {
+  // TODO<developer> set these environment variables
+  private static final String PROJECT_ID = System.getenv("GCP_PROJECT");
+  private static final String TRANSLATE_TOPIC_NAME = System.getenv("TRANSLATE_TOPIC");
+  private static final String[] TO_LANGS = System.getenv("TO_LANG").split(",");
+
+  private static final Logger LOGGER = Logger.getLogger(OcrProcessImage.class.getName());
+  private static final String LOCATION_NAME = LocationName.of(PROJECT_ID, "global").toString();
+  private static Publisher publisher;
+  private static final Gson gson = new Gson();
+
+  public OcrProcessImage() throws IOException {
+    publisher = Publisher.newBuilder(
+        ProjectTopicName.of(PROJECT_ID, TRANSLATE_TOPIC_NAME)).build();
+  }
+
+  @Override
+  public void accept(GcsEvent gcsEvent, Context context) throws RuntimeException {
+
+    // Validate parameters
+    String bucket = gcsEvent.getBucket();
+    if (bucket == null) {
+      throw new IllegalArgumentException("Missing bucket parameter");
+    }
+    String filename = gcsEvent.getName();
+    if (filename == null) {
+      throw new IllegalArgumentException("Missing name parameter");
+    }
+
+    detectText(bucket, filename);
+  }
+  // [END functions_ocr_process]
+
+  // [START functions_ocr_detect]
+  private void detectText(String bucket, String filename) throws RuntimeException {
+    LOGGER.info("Looking for text in image " + filename);
+
+    List<AnnotateImageRequest> visionRequests = new ArrayList<>();
+    String gcsPath = String.format("gs://%s/%s", bucket, filename);
+
+    ImageSource imgSource = ImageSource.newBuilder().setGcsImageUri(gcsPath).build();
+    Image img = Image.newBuilder().setSource(imgSource).build();
+
+    Feature textFeature = Feature.newBuilder().setType(Feature.Type.TEXT_DETECTION).build();
+    AnnotateImageRequest visionRequest =
+        AnnotateImageRequest.newBuilder().addFeatures(textFeature).setImage(img).build();
+    visionRequests.add(visionRequest);
+
+    // Detect text in an image using the Cloud Vision API
+    AnnotateImageResponse visionResponse;
+    try (ImageAnnotatorClient client = ImageAnnotatorClient.create()) {
+      visionResponse = client.batchAnnotateImages(visionRequests).getResponses(0);
+      if (visionResponse.hasError()) {
+        LOGGER.severe("Error detecting text: " + visionResponse.getError().getMessage());
+        return;
+      }
+    } catch (IOException e) {
+      // Cast to RuntimeException
+      throw new RuntimeException(e);
+    }
```

Comment:
Why do we do this?

Reply:
Just because you use a try w/ resources doesn't mean you need to ever catch anything.

Reply:
The Base `BackgroundFunction` type in the Functions Framework can **only** throw subclasses of `RuntimeException`, which these are not.

(IntelliJ requires the exception to be addressed in a `catch` block - even if that block merely ignores the error.)

Reply:
https://google.github.io/styleguide/javaguide.html#s6.2-caught-exceptions
Also see Effective Java v3 Exceptions
* Item 69: Use exceptions only for exceptional conditions 
* Item 70: Use checked exceptions for recoverable conditions and runtime exceptions for programming errors
* Item 71: Avoid unnecessary use of checked exceptions 
* Item 72: Favor the use of standard exceptions 
* Item 73: Throw exceptions appropriate to the abstraction
* Item 74: Document all exceptions thrown by each method 
* Item 77: Don’t ignore exceptions

Reply:
What if you just say `throws IOException` ?

Reply:
Ah - so (some of) these seem more like "checked" exceptions.

Would it be better to log them and exit the function?

Reply:
Please log over `RuntimeException` everywhere.